README.markdown
Tables
======

[![Hackage](https://img.shields.io/hackage/v/tables.svg)](https://hackage.haskell.org/package/tables) [![Build Status](https://secure.travis-ci.org/ekmett/tables.png?branch=master)](http://travis-ci.org/ekmett/tables)

This package provides simple in memory data tables with multiple indices.

Examples
--------

So if we load `examples/Foo.hs` into `ghci`, we start with:

```haskell
>>> test
fromList [ Foo {fooId = 1, fooBar = "One", fooBaz = 1.0}
         , Foo {fooId = 2, fooBar = "Two", fooBaz = 2.0}
         , Foo {fooId = 3, fooBar = "Three", fooBaz = 3.0}
         , Foo {fooId = 4, fooBar = "Four", fooBaz = 4.0}
         , Foo {fooId = 5, fooBar = "Five", fooBaz = 5.0} ]
```

We use uppercase constructor names to match on built-in keys

```haskell
>>> test ^. with FooId (<) 3
fromList [ Foo {fooId = 1, fooBar = "One", fooBaz = 1.0}
         , Foo {fooId = 2, fooBar = "Two", fooBaz = 2.0} ]
```

Then we can use any lowercase field accessor (or any other function) to do a non-keyed lookup or filter

```haskell
>>> test ^. with (length . fooBar) (<=) 3
fromList [ Foo {fooId = 1, fooBar = "One", fooBaz = 1.0}
         , Foo {fooId = 2, fooBar = "Two", fooBaz = 2.0} ]
```

You can delete by assigning to that filtered table:

```haskell
>>> test & with (length . fooBar) (<=) 3 .~ empty
fromList [ Foo {fooId = 3, fooBar = "Three", fooBaz = 3.0}
         , Foo {fooId = 4, fooBar = "Four", fooBaz = 4.0}
         , Foo {fooId = 5, fooBar = "Five", fooBaz = 5.0} ]
```

You can edit the actual type of the fields if the table is configured to allow it:

```haskell
>>> test & rows.fooBar_ %~ length
fromList [ Foo {fooId = 1, fooBar = 3, fooBaz = 1.0}
         , Foo {fooId = 2, fooBar = 3, fooBaz = 2.0}
         , Foo {fooId = 3, fooBar = 5, fooBaz = 3.0}
         , Foo {fooId = 4, fooBar = 4, fooBaz = 4.0}
         , Foo {fooId = 5, fooBar = 4, fooBaz = 5.0} ]
```

If you edit multiple fields, the edits all take place at the same time. so we can offset or swap a bunch of keys:

```haskell
>>> test & with FooId (>=) 2.rows.fooId_ +~ 1
fromList [ Foo {fooId = 1, fooBar = "One", fooBaz = 1.0}
         , Foo {fooId = 3, fooBar = "Two", fooBaz = 2.0}
         , Foo {fooId = 4, fooBar = "Three", fooBaz = 3.0}
         , Foo {fooId = 5, fooBar = "Four", fooBaz = 4.0}
         , Foo {fooId = 6, fooBar = "Five", fooBaz = 5.0} ]
```

We can do grouping by arbitrary functions or fields similarly

```haskell
>>> test ^@.. group (length.fooBar)
[ (3, fromList [ Foo {fooId = 1, fooBar = "One", fooBaz = 1.0}
               , Foo {fooId = 2, fooBar = "Two", fooBaz = 2.0} ])
, (4, fromList [ Foo {fooId = 4, fooBar = "Four", fooBaz = 4.0}
               , Foo {fooId = 5, fooBar = "Five", fooBaz = 5.0} ])
, (5, fromList [Foo {fooId = 3, fooBar = "Three", fooBaz = 3.0} ])
]
```

Contact Information
-------------------

Contributions and bug reports are welcome!

Please feel free to contact me through github or on the #haskell or #haskell-lens IRC channels on irc.freenode.net.

-Edward Kmett


CHANGELOG.markdown
0.4.1.1
----
* GHC 7.10 compatibility

0.4.1
----

0.4
----

0.3.1
-----
* Improved `Setup.lhs` to work with more versions of `cabal`.
* Internal compatibility with changes in `lens` 3.10 to `anon`.

0.3
---
* Instances for `SafeCopy`, `Binary` and `Cereal`
* Added `insert'`

0.2
---
* Generalized Autoincrement
* Better type inference for key types
* More aggressive inlining

0.1
---
* Initial public release
* Repository initialized


examples/Graph.hs
{-# Language TemplateHaskell #-}
{-# Language TypeFamilies #-}
{-# Language GADTs #-}
{-# Language GeneralizedNewtypeDeriving #-}
{-# Language RankNTypes #-}
-- base
import Control.Applicative
-- containers
import Data.Set (Set)
import qualified Data.Set as S
-- lens
import Control.Lens
-- tables
import Data.Table

data Node = Node
  { _name :: Int
  , _arcs :: Set Int
  } deriving (Show)
makeLenses ''Node

instance Tabular Node where
  type PKT Node = Int
  data Key k Node b where
    Key  :: Key Primary Node Int
    Arcs :: Key Inverted Node (Set Int)
  data Tab Node i = NTab (i Primary Int)
    (i Inverted (Set Int))
  fetch Key = _name
  fetch Arcs = _arcs
  primary = Key
  primarily Key r = r
  mkTab f = NTab <$> f Key <*> f Arcs
  forTab (NTab i s) f = NTab <$> f Key i <*> f Arcs s
  ixTab (NTab i _) Key = i
  ixTab (NTab _ s) Arcs = s

main :: IO ()
main = do
  -- Make a lot of graph nodes.
  let t = fromList
        [ Node 1 (S.fromList [1, 2, 3])
        , Node 2 (S.fromList [1, 2, 3])
        , Node 3 (S.fromList [1, 2])
        , Node 4 (S.fromList [5])
        , Node 5 (S.fromList [4])
        , Node 6 (S.fromList [3])
        ]
  -- Print the names of all the nodes pointing to #3.
  print $ t ^.. withAny Arcs [3] . folded . name
  -- Print the names of all the nodes pointing to both #1 and #2.
  print $ t ^.. withAll Arcs [1, 2] . folded . name


examples/Foo.hs
{-# LANGUAGE GADTs #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE DeriveDataTypeable #-}
{-# LANGUAGE LiberalTypeSynonyms #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE FunctionalDependencies #-}
module Foo where

import Control.Applicative hiding (empty)
import Control.Lens
import Data.Data
import Data.Foldable as Foldable
import Data.Function (on)
import Data.Functor.Identity
import Data.List ((\\))
import Data.Map (Map)
import qualified Data.Map as Map
import Data.Monoid
import Data.Table
import Data.Traversable
import qualified Prelude
import Prelude hiding (null)

-- * Example Table

data Foo a = Foo { fooId :: Int, fooBar :: a, fooBaz :: Double }
  deriving (Eq,Ord,Show,Read,Data,Typeable)

makeLensesFor
  [("fooId", "fooId_"), ("fooBar", "fooBar_"), ("fooBaz","fooBaz_")]
  ''Foo

instance Tabular (Foo a) where
  type PKT (Foo a) = Int
  data Key k (Foo a) b where
    FooId  :: Key Primary   (Foo a) Int
    FooBaz :: Key Candidate (Foo a) Double
  data Tab (Foo a) i = FooTab (i Primary Int) (i Candidate Double)

  fetch FooId  = fooId
  fetch FooBaz = fooBaz

  primary = FooId
  primarily FooId r = r

  mkTab f               = FooTab <$> f FooId   <*> f FooBaz
  forTab (FooTab x y) f = FooTab <$> f FooId x <*> f FooBaz y
  ixTab (FooTab x _) FooId  = x
  ixTab (FooTab _ x) FooBaz = x

  autoTab = autoIncrement fooId_

test :: Table (Foo String)
test = [Foo 0 "One" 1.0, Foo 0 "Two" 2.0, Foo 0 "Three" 3.0, Foo 0 "Four" 4.0, Foo 0 "Five" 5.0]^.table
